### PR TITLE
Simplify hybrid evaluation

### LINKED ISSUE #912: Hybrid evaluation of [ fails
e.g.

``` R
library("dplyr")
library("nycflights13") 
flights %>% mutate(airport = airports[, "name"]) 
# Error: not compatible with STRSXP 
```

### LINKED ISSUE #1012: Build variable interpolation into hybrid evaluator
I see a lot of code that does something like this: 

```
f <- function(data, variable, value ){
   filter_( data, interp( ~ var == value, var = as.name(variable) ) )
}
f( iris, "Species", "setosa" )
```

which I find kind of verbose and perhaps not so easy to read. 

I was thinking of something like this: 

```
filter( data, column(variable) == value )
```

`column` would take a string like `"Species"` and substitute it to the associated name, i.e. `Species` so that eventually it is just like we were doing `filter(data, Species == value )`. 

I think it would be fairly easy to do internally, and the code would be IMO much more readable. @hadley what do you think ?

### LINKED ISSUE #1400: Better $ hybrid support
It would be nice if this worked:

``` R
library(boot)
library(dplyr)

df <- bind_rows(
  data_frame(t=1:10, ts="A", dat=1:10),
  data_frame(t=1:10, ts="B", dat=101:110)
)

bsdf <- group_by(df, ts) %>% 
  do(bs = boot(.$dat, R = 10, statistic = function(d, ind) mean(d[ind])))

bsdf %>% mutate(
  mean = bs$t0, # ok 
  var = var(bs$t) # fails
)
```

(which I'm pretty sure is because hybrid $ isn't quite right)

@romainfrancois can you take a look? If you point me to the hybrid implementation of `$` I can also think about improvements.

### LINKED ISSUE #1452: Unclear limitations on the calculations that can be performed in mutate.
I’d like to use `mutate` to create a new column but the calculations are not entirely trivial.  Here is one example illustrating the kind of things I would like to do:

``` R
d %>%
  group_by(tid) %>%
  mutate(before.event = {
    idx <- which(diff(chars) < 0)[1]
     c(rep(T, idx), rep(F, n()-idx))
  })
```

This code produces an error message saying that assignments are forbidden.  So I tried the following work-around:

``` R

f <- function(x) {
  idx <- which(diff(x) < 0)[1]
  c(rep(T, idx), rep(F, n()-idx))
}

d %>%
  group_by(tid) %>%
  mutate(before.event = f(chars))
```

but this also produces an error message:

```
Error: This function should not be called directly
```

This is fairly surprising because I can of course use other functions in mutate, e.g:

``` R
d %>%
  group_by(tid) %>%
  mutate(test = min(chars))
```

I looked in various places for an explanation but couldn’t find one (man pages, GitHub page, introduction to dplyr, …).  There clearly seem to be limitations as to which things can be done and which functions can be used but it is not clear what these limitations are.  This can be of course be a little discouraging for new users.  So I’d say there is at least a documentation issue here.

While this issue could be addressed by improving the documentation, it would of course be much better if there weren’t any limitations in the first place.  I can think of many situations where it would be extremely useful if assignments and user-defined functions were allowed.  I really like dplyr's interface but the ability to use arbitrary code-blocks to carry out calculations in conjunction with `mutate` but also `summarise` would clearly set dplyr apart from alternative solutions like reshape and data.table.

### LINKED ISSUE #1719: Hybrid support for case_when
We should talk about how to implement this - one generic approach (that might also work for `lm()`) is to construct an environment populated with promises that evaluate to the variables in the current group. Then if you call the function in the context of that environment, it should be able to use regular evaluation to get what it needs.

### LINKED ISSUE #2018: mutate error: 'object not found' occurs sometimes when passing variables into functions
I'm finding weird behavior in dplyr mutate when passing variables into functions. Sometimes I'm getting an error that the variable is not found. See below for two examples. In the first I'm calculating the cdf of some values using ecdf. This works fine until I try to subtract 1. In the second example, I'm trying to pass variables into a wilcox.test and getting the same error. Both problems are fixed if I write a wrapper function first. Is this a known issue? Is there something that predicts this behavior?

``` r
#dplyr 0.5.0
library(dplyr)

test_data = data.frame(
  grp = rep(c("A", "B"), each = 4),
  y = rnorm(8), stringsAsFactors = F
)

#this works
test_data %>% group_by(grp) %>%
  mutate(
    cdf = ecdf(y)(y)
  )

test_data %>% group_by(grp) %>%
  mutate(
    mean_y = mean(y) - 1
  )

#this throws an error:  Error: object 'y' not found 
test_data %>% group_by(grp) %>%
  mutate(
    surv = 1 - ecdf(y)(y) 
  )

#but this works
custom_fun = function(input) 1 - ecdf(input)(input)

test_data %>% group_by(grp) %>%
  mutate(
    surv = custom_fun(y)
  )

# example with wilcox.test

test_data2 = data.frame(
  grp = rep(c("A", "B"), each = 4),
  grp2 = rep(c("C", "D"), 4),
  y = rnorm(8), stringsAsFactors = F
)

test_data2 %>% group_by(grp) %>%
  mutate(
    p_value = 1 - wilcox.test(y)$p.value
  )

# Error: object 'y' not found 
test_data2 %>% group_by(grp) %>%
  mutate(
    p_value = wilcox.test(y ~ grp2)$p.value
  )

wilcox_fun = function(outcome, group) wilcox.test(outcome ~ factor(group))$p.value

test_data2 %>% group_by(grp) %>%
  mutate(
    p_value = wilcox_fun(y, grp2)
  )
```

### LINKED ISSUE #2125: dplyr::mutate: invalid subscript type 'closure' when using $ inside expression
I was told that maybe this behaviour is a bug in my [stackoverflow question](http://stackoverflow.com/questions/39494445/why-is-this-construct-not-allowed-in-dplyrmutate):

```
library(dplyr)
df = data.frame(col1=rep(1:5, c(3,3,2,1,1)))
df <- df %>%
     mutate(rle=sequence(rle(col1)$lengths)) 
```

fails with 

```
Error in eval(expr, envir, enclos): invalid subscript type 'closure'
```

Using `mutate(rle=sequence(rle(col1)[["lengths"]])` works so it seems that the problem is the use of `$lengths` vs `[["lenghts"]]`.

### LINKED ISSUE #2160: Use group_indices() in mutate() with nse
Can we get `group_indices` to work in `mutate` using non-standard evaluation? 

Often we want to add the output of `group_indices` back to the dataframe for further manipulation. It is possible with se, but it's a lot of typing and could be more novice-friendly.  

This would be good if it would work:

```
mtcars %>%  mutate(group = group_indices(cyl))
Error: no applicable method for 'group_indices_' applied to an object of class "c('double', 'numeric')"
```

Instead we have to do this, using the se version:

```
  mtcars %>% mutate(group = group_indices_(., .dots="cyl"))
```

```
 mpg cyl  disp  hp drat    wt  qsec vs am gear carb group
1  21.0   6 160.0 110 3.90 2.620 16.46  0  1    4    4     2
2  21.0   6 160.0 110 3.90 2.875 17.02  0  1    4    4     2
3  22.8   4 108.0  93 3.85 2.320 18.61  1  1    4    1     1
4  21.4   6 258.0 110 3.08 3.215 19.44  1  0    3    1     2
5  18.7   8 360.0 175 3.15 3.440 17.02  0  0    3    2     3
6  18.1   6 225.0 105 2.76 3.460 20.22  1  0    3    1     2
7  14.3   8 360.0 245 3.21 3.570 15.84  0  0    3    4     3
8  24.4   4 146.7  62 3.69 3.190 20.00  1  0    4    2     1
9  22.8   4 140.8  95 3.92 3.150 22.90  1  0    4    2     1
10 19.2   6 167.6 123 3.92 3.440 18.30  1  0    4    4     2
11 17.8   6 167.6 123 3.92 3.440 18.90  1  0    4    4     2
```

### LINKED ISSUE #2223: using a missing column name in a group_by crashes R instead of generating an error
```
library(dplyr)
test.df <- data.frame(ids=rep(1:10, 10), vals=1:100)
test.df %>% group_by(ids) %>% summarize(m=mean(vals))
# ok to this point

test.df %>% group_by(ids) %>% summarize(m=mean(missing.col.name))
# this line causes R to crash for me running it inside of RStudio
```

I'm running R version 3.3.1 on ubuntu 16.04

### LINKED ISSUE #2244: case_when fails on grouped dfs
This might be resolved in https://github.com/hadley/dplyr/pull/2190, haven't checked, but am submitting an issue in case others are getting this error: 

`Error in eval(substitute(expr), envir, enclos) :  incompatible size...` 

when using case_when on a grouped df.

```
# Works, wrapping in a separate mutate b/c of #1719 (I think)
iris %>% 
  mutate(ave_length = mean(Sepal.Length)) %>% 
  mutate(length = case_when(
    .$Sepal.Length < .$ave_length ~ "Small", 
    TRUE ~ "Big"))

# Doesn't work
iris %>% 
  group_by(Species) %>%
  mutate(ave_length = mean(Sepal.Length)) %>% 
  mutate(length = case_when(
    .$Sepal.Length < .$ave_length ~ "Small", 
    TRUE ~ "Big"))
```

I ran into this when I forgot that a tibble was grouped.

SessionInfo
<details>
```
R version 3.3.2 (2016-10-31)
Platform: x86_64-w64-mingw32/x64 (64-bit)
Running under: Windows 7 x64 (build 7601) Service Pack 1

locale:
[1] LC_COLLATE=English_United States.1252 
[2] LC_CTYPE=English_United States.1252   
[3] LC_MONETARY=English_United States.1252
[4] LC_NUMERIC=C                          
[5] LC_TIME=English_United States.1252    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] dplyr_0.5.0

loaded via a namespace (and not attached):
[1] lazyeval_0.2.0 magrittr_1.5   R6_2.2.0       assertthat_0.1 DBI_0.5-1     
[6] tools_3.3.2    tibble_1.2     Rcpp_0.12.7 
```
</details>

### PR DESCRIPTION
**Pull request by krlmlr**
Fixes #912.

Fixes #1012 ~~(but untestable because of hadley/lazyeval#78)~~

Fixes #1400.

Fixes #1452.

Fixes #1719.

Fixes #2018.

Fixes #2125.

Fixes #2160.

Fixes #2223.

Fixes #2244.

### CODE REVIEW

**Comment by hadley** [comment]
For testing, that the hybrid evaluator is actually doing something it might be useful to implement #1691, and also define an `in_hybrid()` function that is defined as `function() FALSE` in R, and but returns `TRUE` in the C implementation.

**Comment by krlmlr** [comment]
Performance suffers, a faster Rcpp_eval() would help:

**Comment by krlmlr** [comment]
@hadley: Much better when we create + populate the environment with the active bindings only once (ad4f698). In the tested scenarios run time was slower by < 50%. We still can look into a faster `Rcpp_eval()`, but that would be orthogonal to this PR. I'll do some more local testing before suggesting a merge.

**Comment by hadley** [comment]
That doesn't look too bad at all.

(Also it looks like every big change is currently with both before and after commits. Probably on the the after label is necessary)

**Comment by krlmlr** [comment]
Thanks. I've thought about showing only "after", but I like having "before" as a way to double-check and resolve potential ambiguities. Please also note that for the level established at 2c095 in the "windowed" plot, no "before" entry is detected before 7d918.

**Comment by krlmlr** [comment]
@hadley: PTAL. I'll release bindrcpp to CRAN soon, but this shouldn't affect the review. Tests now show one warning, which should go with #2298.

NEWS entry:

```md
* Breaking change: The `column()` and `global()` functions have been removed. They were never documented officially.
* Expressions in verbs are now interpreted correctly in many cases that failed before (e.g., use of `$`, `case_when()`, nonstandard evaluation, ...). This is possible, because these expressions are now evaluated in a specially constructed environment that retrieves column data on demand with the help of the `bindrcpp` package (#2190).
```

**Comment by hadley** [comment]
Could you do another benchmark plot please?

Should the news bullets also include mention of the new `.env` and `.data` pronouns? (And if you haven't implemented them yet, can you please do so?)

**Comment by krlmlr** [comment]
@hadley: Will do a benchmark plot. What should happen if the data frame contains columns named `.env` and/or `.data`?

**Comment by hadley** [comment]
@krlmlr I think our definition should win so if you wanted to access a column named `.env` you'd need to do `.data$.env`

**Comment by krlmlr** [comment]
@hadley: Done. Performance is very much comparable with `master`, see plots below. NEWS entries:

* Breaking change: The new `.data` and `.env` environments can be used inside all verbs that operate on data: `.data$column_name` accesses the column `column_name`, whereas `.env$var` accesses the external variable `var`. Columns or external variables named `.data` or `.env` are shadowed, use `.data$...` and/or `.env$...` to access them.
* Breaking change: The `column()` and `global()` functions have been removed. They were never documented officially. Use the new `.data` and `.env` environments instead.
* Expressions in verbs are now interpreted correctly in many cases that failed before (e.g., use of `$`, `case_when()`, nonstandard evaluation, ...). This is possible, because these expressions are now evaluated in a specially constructed environment that retrieves column data on demand with the help of the `bindrcpp` package (#2190).

<details>
  <summary>Updated plots</summary>











</details>

File: tests/testthat/test-hybrid-traverse.R [hunk 1]
```diff
@@ -10,23 +10,456 @@ test_df <- data_frame(
...
+      grouping %>%
+      filter(b < 2) %>%
+      select(-e))
+})
+
+test_that("filter understands .env (#1469)", {
+  b <- 2L
+
+  testthat::expect_equal(
+    filter(
```
Comment (hadley) [inline tests/testthat/test-hybrid-traverse.R]: The formatting of this seems a bit off

**Comment by hadley** [comment]
LGTM

**Comment by krlmlr** [comment]
@hadley: Added minor reformatting tweaks and NEWS, will squash-merge after CI has run. Thanks for the fast reviews.

**Comment by krlmlr** [comment]
@lionel-: This is the core of the new hybrid evaluator: Two new environments, one with the variables, one with `.data` and `.env`. The `GroupedHybridCall` class currently does too many things, splitting in #2315.

**Pull request merged**